File: Javascript/DCP3_serialize_tree.js
class Node 
{
  constructor(val, left = null, right = null) 
  {
    this.val = val;
    this.left = left;
    this.right = right;
  }

  add(node, val)
  {
    if(val < node.val)
    {
      if(node.left == null)
      {
        node.left = new Node(val);
      }

      else
      {
        this.add(node.left, val);
      }
    }

    else
    {
      if(node.right == null)
      {
        node.right = new Node(val);
      }

      else
      {
        this.add(node.right, val);
      }
    }
  }

  breadth_first(node)
  {
    let level_order = [];
    for(let i = 0; i < this.height(node) + 1; i++)
    {
      this.print_level(node, i, level_order);
    }

    return level_order;
  }

  print_level(node, level, level_order)
  {
    if(node != null)
    {
      if(level == 0)
      {
        level_order.push(node.val);
      }

      else
      {
        this.print_level(node.left, level - 1, level_order);
        this.print_level(node.right, level - 1, level_order);
      }
    }
  }

  height(node)
  {
    if(node == null)
    {
      return -1;
    }

    else
    {
      return 1 + Math.max(this.height(node.left), this.height(node.right));
    }
  }

  serialize(node)
  {
    return this.breadth_first(node).join('');
  }

  deserialize(str)
  {
    let new_node = new Node(parseInt(str[0]));
    str.split('').forEach(letter =>
    {
      if(letter != str[0])
      {
        this.add(new_node, parseInt(letter));
      }
    });
    return new_node;
  }

}

n = new Node(3);
n.add(n, 1);
n.add(n, 5);
n.add(n, 0);
n.add(n, 2);
n.add(n, 4);
n.add(n, 6);
console.log(n.breadth_first(n)); // [3, 1, 5, 0, 2, 4, 6]
console.log(n.breadth_first(n.deserialize(n.serialize(n)))); // also [3, 1, 5, 0, 2, 4, 6]

File: Python/DCP3_serialize_tree.py
class Node:
    def __init__(self, val, left=None, right=None):
      self.val = val
      self.left = left
      self.right = right

    def add(self, node, val):
      if(val < node.val):
        if(node.left is None):
          node.left = Node(val)
        else:
          self.add(node.left, val)
      else:
        if(node.right is None):
          node.right = Node(val)
        else:
          self.add(node.right, val)

    def depth_first(self, node):
      level_order_result = []
      for i in range(0, self.height(node) + 1):
        self.print_level(node, i, level_order_result)
      return level_order_result

    def print_level(self, node, level, level_order_result):
      if(level is 0 and node != None):
        level_order_result.append(node.val)
      else:
        if(node.left != None):
          self.print_level(node.left, level - 1, level_order_result)
        if(node.right != None):
          self.print_level(node.right, level - 1, level_order_result)

    def height(self, node):
      if node is None:
        return -1
      else:
        return 1 + max(self.height(node.left), self.height(node.right))

    def serialize(self, root):
      level_order_result = self.depth_first(root)
      serialization = []
      for elem in level_order_result:
        serialization.append(str(elem))
      return ''.join(serialization)

    def deserialize(self, str):
      new_node = Node(int(str[0]))
      for i in range(1, len(str)):
        new_node.add(new_node, int(str[i]))
      return new_node

node = Node(3)
node.add(node, 1)
node.add(node, 0)
node.add(node, 2)
node.add(node, 5)
node.add(node, 4)
node.add(node, 6)
print(node.depth_first(node)) # [3, 1, 5, 0, 2, 4, 6]
print(node.depth_first(node.deserialize(node.serialize(node)))) # also [3, 1, 5, 0, 2, 4, 6]
